### DISCUSSION TITLE
Source Code for flash.bin and ctrl_*.bin files

### DISCUSSION THREAD

**Discussion by LarryMallard** [slu4coder/Minimal-64x4-Home-Computer#34] [Q&A]
Super cool project. I'm trying to learn more. How does one go about creating their own flash.bin and ctrl_*.bin files?
Is the source code for these available with instruction to converting to bin format?

**Comment by slu4coder** [comment]
Hi and welcome to the Minimal 64x4. You can create your own ctrl_*.bin files from the microcode definition given in the repo. You'd define the control signals "in order of qpearance" starting with

#define EO                0b000000000000000000000001
#define EC                0b000000000000000000000010
#define ES                0b000000000000000000000100
...

and then build the content as an array with

uint32_t mCtrlROM[0x4000]
	{
		#include "microcode_def.csv"
		#include "microcode_rom.csv"
	};

Final step would be to save this data to three separate files containing the LSB, MSB, HSB parts of the data, i.e. bits 0-7, bits 8-15, bits 16-23.

I cannot recommend to change the control ROMs unless you know exactly what you're doing.

flash.bin can be build from the assembled version of the file "os.asm" given in the repo. flash.bin contains the operating system as well as all files stored on the FLASH at a later time. The version I provide in the repo not only contains the OS but almost all programs available for the Minimal 64x4.

Hope this helps. Cheers!

**Reply by garo** [reply]
I'm curious to understand why you have kept the source code closed? If you really would want people to start learning from this, you should seriously consider releasing these with an open license.

**Reply by slu4coder** [reply]
Hi garo, in my YT videos I've shown how to simulate a TTL CPU from scratch - no information about this CPU design is closed or hidden here. Personally, I see no point in bloating the ideas I consider worthwhile sharing with 6000+ lines of OpenGL-related stuff.

Cheers!
